code/game/objects/items/weapons/RMS_vr.dm

```diff
@@ -0,0 +1,231 @@
+#define RMS_STEEL "Steel"
+#define RMS_GLASS "Glass"
+#define RMS_BATT "Battery"
+#define RMS_RAND "Random Material"
+
+/obj/item/weapon/rms
+	name = "Rapid Material Synthesizer"
+	desc = "A tool that converts battery charge to materials."
+	icon = 'icons/obj/tools_vr.dmi'
+	icon_state = "rms"
+	item_state = "rms"
+	item_icons = list(
+		slot_l_hand_str = 'icons/mob/items/lefthand_vr.dmi',
+		slot_r_hand_str = 'icons/mob/items/righthand_vr.dmi',
+	)
+	force = 7
+	throwforce = 10
+	throw_speed = 1
+	throw_range = 5
+	w_class = ITEMSIZE_NORMAL
+	origin_tech = list(TECH_ENGINEERING = 3, TECH_MATERIAL = 3)
+	matter = list(DEFAULT_WALL_MATERIAL = 5000)
+	preserve_item = FALSE
+
+	var/mode_index = 1 //start at steel creation
+	var/list/modes = list(RMS_STEEL, RMS_GLASS, RMS_BATT, RMS_RAND)
+	var/stored_charge = 0
+	var/max_charge = 1000000 //large storage, equivalent to 50 steel sheets
+	var/charge_cost = 20000
+	var/charge_cost_r = 50000
+	var/charge_cost_m = 100000
+	var/charge_stage = 0
+	var/emagged = 0
+	var/datum/effect/effect/system/spark_spread/spark_system
+
+/obj/item/weapon/rms/Initialize()
+	. = ..()
+	src.spark_system = new /datum/effect/effect/system/spark_spread
+	spark_system.set_up(5, 0, src)
+	spark_system.attach(src)
+	update_icon()
+
+/obj/item/weapon/pipe_dispenser/Destroy()
+	qdel_null(spark_system)
+	return ..()
+
+/obj/item/weapon/rms/update_icon()
+	charge_stage = round((stored_charge/max_charge)*4)
+	if(charge_stage >= 4)
+		charge_stage = 4
+	add_overlay("rms_charge[charge_stage]")
+
+/obj/item/weapon/rms/examine(mob/user)
+	. = ..()
+	. += display_resources()
+
+/obj/item/weapon/rms/proc/display_resources()
+	return "It currently holds [round(stored_charge/1000)]/[max_charge/1000] kW charge."
+
+/obj/item/weapon/rms/proc/drain_battery(user, battery)
+	var/obj/item/weapon/cell.C = battery
+	if(stored_charge == max_charge)
+		to_chat(user, "<span class='notice'>The Rapid Material Synthesizer is full on charge!.</span>")
+		update_icon()
+		return
+	if(C.charge == 0)
+		to_chat(user, "<span class='notice'>The battery has no charge.</span>")
+		update_icon()
+		return
+	else
+		playsound(get_turf(src), 'sound/machines/click.ogg', 50, 1)
+		if(do_after(user, 2,target = C))
+			stored_charge += C.charge
+			C.charge = 0
+			C.update_icon()
+			to_chat(user, "<span class='notice'>You drain [C].</span>")
+			if(stored_charge > max_charge)
+				stored_charge = max_charge
```

Comment:
`stored_charge = CLAMP(stored_charge, 0, max_charge)`
And might want to do it above the update_icon call because that's mechanically important to the thing, but see my other comment about that